File: training-data/php/word_search_grid.php
<?php

function exists(array $board, string $word): bool
{
    $rows = count($board);
    $cols = count($board[0]);
    $dfs = function (int $r, int $c, int $i) use (&$dfs, &$board, $word, $rows, $cols): bool {
        if ($i === strlen($word)) {
            return true;
        }
        if ($r < 0 || $c < 0 || $r >= $rows || $c >= $cols || $board[$r][$c] !== $word[$i]) {
            return false;
        }
        $saved = $board[$r][$c];
        $board[$r][$c] = '#';
        $found = $dfs($r + 1, $c, $i + 1) || $dfs($r - 1, $c, $i + 1) || $dfs($r, $c + 1, $i + 1) || $dfs($r, $c - 1, $i + 1);
        $board[$r][$c] = $saved;
        return $found;
    };
    for ($r = 0; $r < $rows; $r++) {
        for ($c = 0; $c < $cols; $c++) {
            if ($dfs($r, $c, 0)) {
                return true;
            }
        }
    }
    return false;
}

$board = array_map('str_split', ['ABCE', 'SFCS', 'ADEE']);
foreach (['ABCCED', 'SEE', 'ABCB'] as $w) {
    echo $w, ': ', exists($board, $w) ? 'yes' : 'no', PHP_EOL;
}

File: training-data/c/word_search_grid.c
#include <stdbool.h>
#include <stdio.h>

#define R 3
#define C 4

static bool dfs(char b[R][C + 1], const char *w, int r, int c) {
    if (!*w) return true;
    if (r < 0 || r >= R || c < 0 || c >= C || b[r][c] != *w) return false;
    char tmp = b[r][c];
    b[r][c] = '#';
    bool found = dfs(b, w + 1, r + 1, c) || dfs(b, w + 1, r - 1, c) || dfs(b, w + 1, r, c + 1) || dfs(b, w + 1, r, c - 1);
    b[r][c] = tmp;
    return found;
}

bool exists(char b[R][C + 1], const char *word) {
    for (int r = 0; r < R; r++)
        for (int c = 0; c < C; c++)
            if (dfs(b, word, r, c)) return true;
    return false;
}

int main(void) {
    char b[R][C + 1] = {"ABCE", "SFCS", "ADEE"};
    const char *words[] = {"ABCCED", "SEE", "ABCB"};
    for (int i = 0; i < 3; i++) printf("%s %s\n", words[i], exists(b, words[i]) ? "true" : "false");
    return 0;
}
